sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/Watch.java

```diff
@@ -0,0 +1,811 @@
+
+      Coder<OutputT> outputCoder = getOutputCoder();
+      if (outputCoder == null) {
+        // If a coder was not specified explicitly, infer it from the OutputT type parameter
+        // of the PollFn.
+        TypeDescriptor<?> superDescriptor =
+            TypeDescriptor.of(getPollFn().getClass()).getSupertype(PollFn.class);
+        TypeVariable typeVariable = superDescriptor.getTypeParameter("OutputT");
+        @SuppressWarnings("unchecked")
+        TypeDescriptor<OutputT> descriptor =
+            (TypeDescriptor<OutputT>) superDescriptor.resolveType(typeVariable);
+        try {
+          outputCoder = input.getPipeline().getCoderRegistry().getCoder(descriptor);
+        } catch (CannotProvideCoderException e) {
+          throw new RuntimeException(
+              "Unable to infer coder for OutputT. Specify it explicitly using withOutputCoder().");
+        }
+      }
+
+      return input
+          .apply(ParDo.of(new WatchGrowthFn<>(this, outputCoder)))
+          .setCoder(KvCoder.of(input.getCoder(), outputCoder));
+    }
+  }
+
+  private static class WatchGrowthFn<InputT, OutputT> extends DoFn<InputT, KV<InputT, OutputT>> {
+    private final Watch.Growth<InputT, OutputT> spec;
+    private final Coder<OutputT> outputCoder;
+
+    private WatchGrowthFn(Growth<InputT, OutputT> spec, Coder<OutputT> outputCoder) {
+      this.spec = spec;
+      this.outputCoder = outputCoder;
+    }
+
+    @ProcessElement
+    public ProcessContinuation process(ProcessContext c, final GrowthTracker<OutputT> tracker) {
+      if (!tracker.hasPending() && !tracker.currentRestriction().isOutputFinal) {
+        tracker.addNewAsPending(
+            spec.getPollFn().apply(TimestampedValue.of(c.element(), c.timestamp())));
+      }
+      while (tracker.hasPending()) {
+        c.updateWatermark(tracker.getWatermark());
+
+        TimestampedValue<OutputT> nextPending = tracker.tryClaimNextPending();
+        if (nextPending == null) {
+          return stop();
+        }
+        c.outputWithTimestamp(
+            KV.of(c.element(), nextPending.getValue()), nextPending.getTimestamp());
+      }
+      // No more pending outputs - future output will come from more polling,
+      // unless output is final.
+      if (tracker.isOutputFinal()) {
+        return stop();
+      }
+      return resume().withResumeDelay(spec.getPollInterval());
+    }
+
+    @GetInitialRestriction
+    public GrowthState<OutputT> getInitialRestriction(InputT element) {
+      return new GrowthState<>(spec.getTerminationPerInput().forNewInput(Instant.now()));
+    }
+
+    @GetRestrictionCoder
+    public Coder<GrowthState<OutputT>> getRestrictionCoder() {
+      return GrowthStateCoder.of(outputCoder);
+    }
+  }
+
+  @VisibleForTesting
+  static class GrowthState<OutputT>
+      implements HasDefaultTracker<GrowthState<OutputT>, GrowthTracker<OutputT>> {
+    private final Set<OutputT> completed;
```

Comment:
Perhaps this should be a compact set of sufficiently unique hashes, and PollFn should return merely an Iterable rather than List, or should have a push-like interface, to support extremely large poll results (e.g. a few dozen million outputs). Then the state can still be pretty compact - a set of hashes doesn't take up much space, and we'll just need to not add excessive new outputs to "pending" (e.g. stop adding new pending outputs after having added 100K of them, and in that case rely on the next poll - which then will need to be done unconditionally, regardless of whether the current poll result was final or not).